File: src/main/java/ru/citeck/ecos/rest/AuthenticationService.java
package ru.citeck.ecos.rest;

import com.intellij.openapi.components.Service;
import com.intellij.openapi.project.Project;
import ru.citeck.ecos.settings.EcosServer;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Service
public final class AuthenticationService {

    public static String BASIC_METHOD = "Basic";
    public static String OAUTH2_METHOD = "OAuth2";

    public static String GRANT_TYPE_PASSWORD = "password";
    public static String GRANT_TYPE_CLIENT_CREDENTIALS = "client_credentials";

    public static Collection<String> AUTH_METHODS = List.of(
            BASIC_METHOD, OAUTH2_METHOD
    );

    public static Collection<String> GRANT_TYPES = List.of(
            GRANT_TYPE_PASSWORD, GRANT_TYPE_CLIENT_CREDENTIALS
    );

    private final Map<EcosServer, Authenticator> authenticators = new ConcurrentHashMap<>();

    public Authenticator getAuthenticator(EcosServer ecosServer, Project project) {
        Authenticator authenticator = authenticators.get(ecosServer);
        if (authenticator == null) {
            if (BASIC_METHOD.equals(ecosServer.getAuthMethod())) {
                authenticator = new BasicAuthenticator(ecosServer, project);
            } else if (OAUTH2_METHOD.equals(ecosServer.getAuthMethod())) {
                authenticator = new OAuth2Authenticator(ecosServer, project);
            }
            authenticators.put(ecosServer, authenticator);
        }
        return authenticator;
    }

}

File: src/main/kotlin/ru/citeck/idea/authentication/AuthenticationService.kt
package ru.citeck.idea.authentication

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import ru.citeck.idea.authentication.basic.BasicAuthenticator
import ru.citeck.idea.authentication.oidc.OidcAuthenticator
import ru.citeck.idea.http.CiteckHttpClient
import ru.citeck.idea.settings.servers.CiteckServer
import java.util.concurrent.ConcurrentHashMap

@Service
class AuthenticationService {

    companion object {
        private val log = Logger.getInstance(AuthenticationService::class.java)

        @JvmStatic
        fun getInstance(): AuthenticationService {
            return ApplicationManager.getApplication().getService(AuthenticationService::class.java)
        }
    }

    private val authenticators: MutableMap<CiteckServer, Authenticator> = ConcurrentHashMap()

    private val httpClient = CiteckHttpClient()

    fun getAuthenticator(citeckServer: CiteckServer): Authenticator {
        var authenticator = authenticators[citeckServer]
        if (authenticator == null) {
            val eisConfig = getServerEisConfig(citeckServer.host)
            authenticator = if (eisConfig == null || eisConfig.eisId == "EIS_ID") {
                BasicAuthenticator(citeckServer)
            } else {
                OidcAuthenticator(citeckServer, eisConfig)
            }
            authenticators[citeckServer] = authenticator
        }
        return authenticator
    }

    private fun getServerEisConfig(url: String): ServerEisConfig? {
        return try {
            httpClient.newRequest()
                .uri("$url/eis.json")
                .getForJson(ServerEisConfig::class.java)
                .get()
        } catch (e: Exception) {
            log.warn("Failed to load EIS config from $url/eis.json, falling back to BasicAuthenticator", e)
            null
        }
    }
}
